x.payment_batching/payment_batching.md

```diff
@@ -0,0 +1,307 @@
+  "bc1qsul9emtnz0kks939egx2ssa6xnjpsvgwq9chrw": 0.3
+}'
+```
+
+<!-- for max standard tx size: src/policy/policy.h:static const unsigned int MAX_STANDARD_TX_WEIGHT = 400000; -->
+
+If using your own implementation, you are probably already creating
+transactions with two outputs in most cases, so it should be easy to add
+support for additional outputs.  The only notable consideration is that
+Bitcoin Core nodes (and most other nodes) will refuse to accept or relay
+transactions over 100,000 vbytes, so you should not attempt to send
+batched payments larger than this.
+
+## Combining batching with other opt-in scaling techniques
+
+Payment batching works well with all the techniques described in this
+guide, but some combinations are especially notable:
+
+### Changeless transactions
+
+Changeless transactions save 31 to 43 vbytes (for typical transaction
+templates) by closely matching the total value of the inputs and outputs
+in the transaction, plus fees, so that there's no need to return any
+change back to the spender's wallet.  This is also a perfectly efficient
+way to consolidate inputs as you decrease the total number of your
+inputs by making a payment you would've made anyway.
+
+Changeless transactions always decrease overall transaction size, but
+because the efficiency of payment batching comes from spreading the
+fixed costs of the transaction across multiple receivers, the benefit
+of changeless transactions is weakened (in percentage terms) the more
+you are able to batch.  The following plot compares the best-case
+situation from earlier in this chapter, which included a change output,
+to an otherwise-identical changeless transaction.
+
+![Plot of normal best case against a changeless best
+case](img/p2wpkh-batching-changeless.png)
+
+Although the benefit may be small in percentage terms, the cost of
+implementing changeless transactions may also be small.  Your goal is to
+find a single input you own whose value is equal to or slightly larger
+than the total value of the outputs you want to send.  A simple
+algorithm would be to wait until you have a minimum number of payments
+to batch and check to see if you have an input that is equal to or
+slightly larger than that.  If not, wait until you have another payment
+to send, combine that with the earlier payments, and try again.  Keep
+trying until you reach a limit (e.g. it's time to send the oldest
+payment) or the aggregate output value is larger than your largest
+input.
+
+For more information about changeless transactions, including exactly
+what "slightly larger than the output value" means, see the
+corresponding section in the chapter about [coin selection
+strategies][].
+
+### Replace-by-Fee (RBF)
+
+RBF is typically promoted as a strategy for increasing the fees paid in
+a transaction so that it confirms sooner, but it also allows for adding
+outputs to a transaction, making it suitable for adding even more
+payments to a previously-sent transaction.
```

Comment:
How much risk is there that when fee bumping TX1 (which includes payment P1), with TX2 (w/ P1 and P2), that TX1 gets mined instead of TX2 (either due to a timing issue or bizarre miner behavior choosing the lower fee TX). The service would need to monitor this and realize P2 needs to be resent. Are there other reasons the service would already be doing such monitoring or would this be an additional burden added if using RBF combined with batching?

Reply:
> How much risk is there that when fee bumping TX1 (which includes payment P1), with TX2 (w/ P1 and P2), that TX1 gets mined instead of TX2 (either due to a timing issue or bizarre miner behavior choosing the lower fee TX).

If txes take 15 seconds to propagate to miners, the risk is about 2.5%:

```
In [1]: from math import exp

In [2]: 1-exp(-15/600)
Out[2]: 0.024690087971667385
```

I'll add a note here about the need to specifically monitor for which payments were confirmed.  Good catch, thanks!

Reply:
Oh, goodness.  While writing this, I just realized that it complicates things significantly.  If you send tx0 with {p1,p2}, replace it with tx0' with {p1,p2,p3}, have tx0 confirmed, and finally send tx1 with {p3}---you have to make sure tx1 spends the change of tx0 or a reorg could allow confirmation of both tx0' and tx1.

Maybe this should be removed too.  I'm not sure the gain is worth the complication.  Also, ISTR that Electrum has this behavior implemented; I wonder whether they handle it safely.

Reply:
Very good discovery! FYI I've reached out to Electrum on #electrum asking how they implemented RBF batching in 3.3.

Reply:
ThomasV confirmed on #electrum that Electrum does not automatically create and send tx1 so would not fall into the trap identified.